### PR TITLE
WIP, Don't merge! Clean up the epoll code using helper

### PR DESCRIPTION
**Pull request by alexlarsson**
Reviewing the attach patch in https://github.com/kubernetes-incubator/cri-o/pull/546 its clear that the current hardcoding of the poll loop for epoll doesn't scale.

This PR (based on the latest version of the attach branch) contains a minimal helper for cleaning this up as one approach to this, to start the discussion of how this should look.

Another alternative is to use the glib main loop,which has very nice features for polling fds, including integration of signals and whatnot. However, that does currently use poll() as the backend, not epoll. I don't think that epoll has a practical advantage in the current code, because we're really only handling a handful of fds, so personally I would prefer using the existing tested code in glib. However, maybe there is a particular reason that epoll was chosen?

### CODE REVIEW

**Comment by alexlarsson** [comment]
Needs signed-by, but i can add this later. More interesting is the discussion about what approach to take here.

**Comment by mrunalp** [comment]
This PR looks nice for cleaning up the epoll code. We intended to move from bare epoll to libuv but it doesn't handle the subreaper/signal fds/pipes as we would have wanted so I just continued to build on top of epoll. We only ever expect a few clients and the worst case is a lot of data back and forth over the connection with a single client. We should be fine handling that with glib/poll/level triggered, right?

**Comment by alexlarsson** [comment]
@mrunalp That should work fine. I can make an alternative PR using glib so we can compare.

**Comment by mrunalp** [comment]
@alexlarsson Sure, sounds good. Thanks! Meanwhile, I will chase down test failures in jenkins CI with my current PR.

File: conmon/conmon.c [hunk 1]
```diff
@@ -743,6 +904,72 @@ int main(int argc, char *argv[])
...
+		/*
+		 * We make the socket non-blocking to avoid a race where client aborts connection
+		 * before the server gets a chance to call accept. In that scenario, the server
+		 * accept blocks till a new client connection comes in.
+		 */
+		afd = socket(AF_UNIX, SOCK_STREAM|SOCK_NONBLOCK|SOCK_CLOEXEC, 0);
+		if (afd == -1)
+			pexit("Failed to create attach socket");
+
+                if (fchmod(afd, 0700))
```
Comment (jawnsy) [inline conmon/conmon.c]: indent on this looks different from the others. I guess we should run the code through `astyle` or whatever?

**Comment by mrunalp** [comment]
I guess we can close this one since we merged the glib main loop PR.

**Pull request closed without merging**